capsules/extra/src/pressure.rs

```diff
@@ -0,0 +1,147 @@
+//!
+//! * `Ok(())`: the callback been successfully been configured.
+//! * `ENOSUPPORT`: Invalid allow_num.
+//! * `NOMEM`: No sufficient memory available.
+//! * `INVAL`: Invalid address of the buffer or other error.
+//!
+//!
+//! ### `command` System Call
+//!
+//! The `command` system call support one argument `cmd` which is used to specify the specific
+//! operation, currently the following cmd's are supported:
+//!
+//! * `0`: check whether the driver exist
+//! * `1`: read the barometer
+//!
+//!
+//! The possible return from the 'command' system call indicates the following:
+//!
+//! * `Ok(())`:    The operation has been successful.
+//! * `BUSY`:      The driver is busy.
+//! * `ENOSUPPORT`: Invalid `cmd`.
+//! * `NOMEM`:     No sufficient memory available.
+//! * `INVAL`:     Invalid address of the buffer or other error.
+//!
+//! Usage
+//! -----
+//!
+//! You need a device that provides the `hil::sensors::PressureDriver` trait.
+//!
+//! ```rust
+//! # use kernel::static_init;
+//!
+//! let pressure = static_init!(
+//!        capsules::temperature::PressureSensor<'static>,
+//!        capsules::temperature::PressureSensor::new(si7021,
+//!                                                 board_kernel.create_grant(&grant_cap)));
+//!
+//! kernel::hil::sensors::PressureDriver::set_client(si7021, pressure);
+//! ```
+
+use core::cell::Cell;
+
+use kernel::grant::{AllowRoCount, AllowRwCount, Grant, UpcallCount};
+use kernel::hil;
+use kernel::hil::sensors::PressureClient;
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::{ErrorCode, ProcessId};
+
+/// Syscall driver number.
+use capsules_core::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Pressure as usize;
+
+#[derive(Default)]
+pub struct App {
+    subscribed: bool,
+}
+
+pub struct PressureSensor<'a> {
+    driver: &'a dyn hil::sensors::PressureDriver<'a>,
+    apps: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    busy: Cell<bool>,
+}
+
+impl<'a> PressureSensor<'a> {
+    pub fn new(
+        driver: &'a dyn hil::sensors::PressureDriver<'a>,
+        apps: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    ) -> PressureSensor<'a> {
+        PressureSensor {
+            driver: driver,
+            apps: apps,
+            busy: Cell::new(false),
+        }
+    }
+
+    fn enqueue_command(&self, processid: ProcessId) -> CommandReturn {
+        self.apps
+            .enter(processid, |app, _| {
+                if !self.busy.get() {
+                    app.subscribed = true;
+                    self.busy.set(true);
+                    let res = self.driver.read_atmospheric_pressure();
+                    if let Ok(err) = ErrorCode::try_from(res) {
+                        CommandReturn::failure(err)
+                    } else {
+                        CommandReturn::success()
+                    }
+                } else {
+                    CommandReturn::failure(ErrorCode::BUSY)
```

Comment:
This needs to be ok. It's fine if the driver is busy taking a reading. The app will still get the callback.

Reply:
Can you clarify what you mean by `This needs to be ok`.

Reply:
`CommandReturn::success()`